icaruscode/CRT/CRTPMTMatchingAna_module.cc

```diff
@@ -54,11 +55,94 @@ using std::map;
     class CRTPMTMatchingAna;
+    struct CRTPMT{
+      double tof;
+      double distance;
+      art::Ptr<sbn::crt::CRTHit> CRTHit;
+    };
+    struct MatchedCRT{
+      //vector of pairs where first is the matched Time of Flight and second is the matched CRTHit
+      std::vector< CRTPMT > entering;
+      std::vector< CRTPMT > exiting;
+      int matchType;
+    }; 
+    struct TrajPoint{
+      double X;
+      double Y;
+      double Z;
+      double T;
+    }; 
   }
 }
 
 using namespace icarus::crt;
 
+bool flashInTime( double const & flashTime, int gateType, double gateDiff ){
+  std::map<unsigned int, double> gateLength{
+    {1, 2.7}, //BNB
+    {2, 10.6}, //NuMI
+    {3, 2.7}, //BNB Offbeam
+    {4, 10.6}, //NuMI Offbeam  
+  };
+
+  //double vetoOffset = 3.5;
+  double activeGate = gateLength[gateType] /*- vetoOffset*/;
+
+  double relFlashTime = flashTime + gateDiff/1000. /*- vetoOffset*/;
+  //std::cout<<"Gate Diff "<<gateDiff/1000<<"  "<<flashTime<<" Ftime+gateDiff "<<flashTime + gateDiff/1000.<<" "<<activeGate<<std::endl;
+  return ( (relFlashTime > 0) && (relFlashTime < activeGate) );
+}
+
+icarus::crt::MatchedCRT CRTHitmatched( const double & flashTime, const double flashpos[3], std::vector< art::Ptr<sbn::crt::CRTHit> > & crtHits, const double & interval ){
+
+    std::vector< icarus::crt::CRTPMT > enteringCRTHits;
+    std::vector< icarus::crt::CRTPMT > exitingCRTHits;
+    
+    int hasCRTHit = 0;
+    int topen=0, topex=0, sideen=0, sideex=0;
+    for( auto const crtHit : crtHits ){
+      double tof = crtHit->ts1_ns/1e3 - flashTime;
+      double distance = sqrt(pow(flashpos[0]-crtHit->x_pos,2)+pow(flashpos[1]-crtHit->y_pos,2)+pow(flashpos[2]-crtHit->z_pos,2));
+      //std::cout<<"TOF "<<tof<<" "<<crtHit->ts1_ns<<" "<<crtHit->plane<<std::endl;
+      if( tof < 0 && abs(tof) < interval ){
+        if(crtHit->plane>36) sideen++;
+        else topen++;
+        CRTPMT m_match={tof, distance, crtHit};
+        enteringCRTHits.push_back( m_match );
+      }
+      else if (tof >=0 && abs(tof) < interval){
+        if(crtHit->plane>36) sideex++;
+        else topex++;
+        CRTPMT m_match={tof, distance, crtHit};
+        exitingCRTHits.push_back( m_match );
+      }
+    }
+    if (topen == 0 && sideen == 0 && topex == 0 && sideex == 0 ) hasCRTHit = 0;
+    else if (topen == 1 && sideen == 0 && topex == 0 && sideex == 0 ) hasCRTHit = 1;
+    else if (topen == 0 && sideen == 1 && topex == 0 && sideex == 0 ) hasCRTHit = 2;
+    else if (topen == 1 && sideen == 0 && topex == 0 && sideex == 1 ) hasCRTHit = 3; 
+    else if (topen == 0 && sideen == 0 && topex == 1 && sideex == 0 ) hasCRTHit = 4;
+    else if (topen == 0 && sideen == 0 && topex == 0 && sideex == 1 ) hasCRTHit = 5;
+    else if (topen >= 1 && sideen >= 1 && topex == 0 && sideex == 0 ) hasCRTHit = 6;
+    else if (topen >= 1 && sideen >= 1 && topex == 0 && sideex >= 1 ) hasCRTHit = 7;
+    else hasCRTHit = 8;
+
+    // hasCRTHit = 0, no matched CRT
```

Comment:
Perhaps hasCRTHit could be made an enum for better code clarity?

Reply:
yes it does. It was on my to do list

Reply:
done, thank you for the reminder!